Add Panda Chaika Extension

KenjieDec opened the pull request:
Closes #1821

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

KenjieDec committed:
Add Panda Chaika Extension

- Add "Panda Chaika" extension from panda.chaika.moe

AwkwardPeak7 commented on src/all/pandachaika/build.gradle:
```diff
@@ -0,0 +1,8 @@
+    isNsfw = true
+}
+
+apply from: "$rootDir/common.gradle"
```
```suggestion
apply from: "$rootDir/common.gradle"

```
newline

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,303 @@
+        val publishers = tags.filter { it.startsWith("publisher:") }.emptyToNull()?.joinToString { it.substringAfter(":").replace("_", " ") }
+        val male = tags.filter { it.startsWith("male:") }.emptyToNull()?.joinToString { it.substringAfter(":").replace("_", " ") }
+        val female = tags.filter { it.startsWith("female:") }.emptyToNull()?.joinToString { it.substringAfter(":").replace("_", " ") }
+        val others = tags.filter { it.startsWith("other:") || (!it.contains("female:") && !it.contains("male:") && !it.contains("artist:") && !it.contains("publisher:") && !it.contains("group:") && !it.contains("parody:")) }.emptyToNull()?.joinToString { it.substringAfter(":").replace("_", " ") }
```
use something like this:
```
listOf("female:", "male:", "artist:", "publisher:", "group:", "parody:").none { substring -> it.contains(substring) }
```

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,303 @@
+
+    // Details
+    private fun Archive.toSManga() = SManga.create().apply {
+        val groups = tags.filter { it.startsWith("group:") }.emptyToNull()?.joinToString { it.substringAfter(":").replace("_", " ") }
```
you can make a fucktion out of this

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,303 @@
+        var type = filename.substringAfterLast('.').lowercase()
+        type = if (type == "jpg") "jpeg" else type
+
+        return GET("https://127.0.0.1/?images/$type;base64,$imageUrl", headers)
```
move page fetching from zip to an interceptor

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,303 @@
+
+    // Pages
+    override fun pageListParse(response: Response): List<Page> {
+        fun List<String>.sort() = this.sortedBy { it.filter(Char::isDigit).toIntOrNull() }
```
would this perform a natural sort on filenames?

KenjieDec replied:
It should sort the list based on the numbers in their filenames, as most filenames usually include page numbers.

KenjieDec replied:
They also have a website section that specifies the file index alongside the filename, like [here](https://panda.chaika.moe/archive/53071/image-data-list/). 
It's possible to use that to sort the filenames, but the current sort is simpler.

AwkwardPeak7 replied:
I meant sorting like 1, 2, 3... instead of 10, 1, 2 ... like how normally computers sort. but if it works correctly then it's ok. 

also what if there are other numbers in the file name...?

AwkwardPeak7 replied:
![image](https://github.com/keiyoushi/extensions-source/assets/48650614/9c545ab4-0c24-4f10-aa0b-883b404911f5)
I belive this kind of number will mess with sorting? [archive](https://panda.chaika.moe/archive/53071/)

you can try what app uses
https://github.com/mihonapp/mihon/blob/main/app/src/main/java/eu/kanade/tachiyomi/ui/reader/loader/ArchivePageLoader.kt#L18
https://github.com/mihonapp/mihon/blob/9e2f97eeb8a0f1d1b353dc3e77fb64d69b568674/core/common/src/main/kotlin/eu/kanade/tachiyomi/util/lang/StringExtensions.kt#L37-L40

KenjieDec replied:
Thank you. I will change it to your suggestion 🙏

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,303 @@
+
+        val file = remoteZip.toJson()
+        return fileListing.mapIndexed { index, _ ->
+            Page(index, imageUrl = fileListing[index] + "and then this:" + file)
```
```suggestion
        return fileListing.mapIndexed { index, img ->
            Page(index, imageUrl = img + "#" + file)
```
this will add to the url fragment which can easily be checked and accessed in interceptor

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaikaUtils.kt:
don't use OkhttpClient() directly, use network.cloudflareClient

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaikaUtils.kt:
you can also make one class which you instentiate in the main source class and call it's methods

KenjieDec committed:
Add Support for Zip64

- Add support for Zip64 type of .zip ( large zip [ size/pages ]  ) -> For Example: https://panda.chaika.moe/archive/49406/
- Use Little Endian for All signatures
- Apply AwkwardPeak7's suggestions

KenjieDec committed:
Fix null Genres?

KenjieDec committed:
Fix "null" genre if there's no genre

KenjieDec committed:
Fix mistakes caused by previous commit

Sorry...

KenjieDec committed:
Improve description readability

- Sorry for the commits spam
- Make manga description more readable

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,326 @@
+
+        val fileType = getZipType(url)
+
+        val remoteZip = ZipHandler(url, client, headers, fileType).populate()
```
you are making the same head request twice, you can reduce it by detecting the zip type directly in ZipHandler where you are getting the size

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaikaUtils.kt:
```diff
@@ -0,0 +1,381 @@
+                }",
+            )
+
+        val request = Request.Builder()
```
use GET or POST where needed, and only use Request.Builder() directly when necessary (like HEAD request)

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaikaUtils.kt:
```diff
@@ -0,0 +1,381 @@
+const val END_OF_CENTRAL_DIRECTORY_64_SIGNATURE = 0x06064b50
+const val LOCAL_FILE_HEADER_SIGNATURE = 0x04034b50
+
+data class EndOfCentralDirectory(
```
you can remove `data` and use regular classes

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,326 @@
+        return MangasPage(mangas, hasNextPage)
+    }
+
+    private fun popularManga(hentai: LongArchive) = SManga.create().apply {
```
all the info seems to be availble in `/search/` call so parse everything there, and add `initialized = true` to smanga in order to avoid one extra network call for fetching details

AwkwardPeak7 replied:
also not necessary, but consider moving these functions to dto classes

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    private fun popularManga(hentai: LongArchive) = SManga.create().apply {
+        setUrlWithoutDomain(hentai.url)
```
consider saving id instead of url

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,326 @@
+        return listOf(
+            SChapter.create().apply {
+                name = "Chapter"
+                url = archive.download.substringBefore("/download/")
```
can directly save download url and start parsing the zip directly instead of fetching again in page list parsing. (override fetchPageList for that)

KenjieDec committed:
Fix Broken Filters, Apply Suggestions

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaikaDto.kt:
```diff
@@ -0,0 +1,104 @@
+class LongArchive(
+    private val thumbnail: String,
+    private val title: String,
+    private val tags: List<String>,
```
remove unused

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaikaDto.kt:
```diff
@@ -0,0 +1,104 @@
+import java.util.Locale
+
+@Serializable
+class Archive(
```
like previously mentioned, is it not possible to parsing this whole archive in the `/search/` endpoint? because it looks like all the info is already available there

KenjieDec replied:
Sorry. I can't quite understand. 
Do you mean I should just remove Archive class and use LongArchive, using "/search/xxxx" instead of "?archive=" in mangaDetails and chapterList Request?

AwkwardPeak7 replied:
in popular/latest/search, you call `/search/` which returns list of archives. these archives have all the necessary information (tags/source/category etc etc), so parse everything there (what you do in Archive class). this way (with initialized = true) app won't make additional calls for details.

KenjieDec replied:
Thank you for the suggestion. Sorry I didn't implement that in the previous commits. 
I thought it wouldn't work because I had already tested it before, but it seems like it's just a cache issue on my end.

KenjieDec replied:
Hello, I've changed the code as per your suggestion. Please let me know if it's incorrect. I might have misinterpreted your suggestion.

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaika.kt:
```diff
@@ -0,0 +1,253 @@
+        return listOf(
+            SChapter.create().apply {
+                name = "Chapter"
+                url = "$baseUrl${archive.download}"
```
save relative url (or just the id?)

AwkwardPeak7 commented on src/all/pandachaika/src/eu/kanade/tachiyomi/extension/all/pandachaika/PandaChaikaDto.kt:
```diff
@@ -0,0 +1,104 @@
+                else -> "$bytes B"
+            }
+        }
+        val dateReformat = SimpleDateFormat("EEEE, d MMM yyyy HH:mm (z)", Locale.ENGLISH)
```
oh and put this out of function (file root)

KenjieDec committed:
Apply suggestions

- Apply AwkwardPeak's suggestions
